### ISSUE TITLE
tsnr value error leads to failing camera, and second error due to unassigned variable reference

### ISSUE THREAD

**Issue by akremin** [desihub/desispec#1161]
The calculation of tsnr for one camera of one exposure fails. It succeeded in a prior exposure of the same tile with the same exposure time.

`desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20210224/00077950/frame-z7-00077950.fits --fiberflat /global/cfs/cdirs/desi/spectro/redux/cascades/calibnight/20210224/fiberflatnight-z7-20210224.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20210224/00077950/sky-z7-00077950.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20210224/00077950/fluxcalib-z7-00077950.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20210224/00077950/cframe-z7-00077950.fits --cosmics-nsig 6`

```
INFO:fluxcalibration.py:1371:apply_flux_calibration: starting
INFO:procexp.py:132:main: calculating tsnr
INFO:tsnr.py:265:calc_tsnr2: read NEA file /global/common/software/desi/cori/desiconda/20200801-1.4.0-spec/code/desimodel/0.14.0/data/specpsf/nea/masternea_z7.fits
INFO:tsnr.py:276:calc_tsnr2: read TSNR ensemble files in /global/common/software/desi/cori/desiconda/20200801-1.4.0-spec/code/desimodel/0.14.0/data/tsnr
INFO:tsnr.py:291:calc_tsnr2: RDNOISE    	 2.348 +- 0.168
INFO:tsnr.py:291:calc_tsnr2: NEA        	 3.916 +- 0.055
INFO:tsnr.py:291:calc_tsnr2: ANGPERPIX  	 0.607 +- 0.028
INFO:tsnr.py:291:calc_tsnr2: ANGPERSPECBIN 	 0.800 +- 0.000
ERROR:tsnr.py:210:calc_alpha: tSNR alpha -0.2441 < 0.8
Traceback (most recent call last):
  File "/global/common/software/desi/cori/desiconda/20200801-1.4.0-spec/code/desispec/0.39.1/lib/python3.8/site-packages/desispec-0.39.1-py3.8.egg/desispec/tsnr.py", line 295, in calc_tsnr2
    alpha = calc_alpha(frame, fibermap=frame.fibermap,
  File "/global/common/software/desi/cori/desiconda/20200801-1.4.0-spec/code/desispec/0.39.1/lib/python3.8/site-packages/desispec-0.39.1-py3.8.egg/desispec/tsnr.py", line 211, in calc_alpha
    raise ValueError(msg)
ValueError: tSNR alpha -0.2441 < 0.8

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/global/common/software/desi/cori/desiconda/20200801-1.4.0-spec/code/desispec/0.39.1/bin/desi_process_exposure", line 4, in <module>
    __import__('pkg_resources').run_script('desispec==0.39.1', 'desi_process_exposure')
  File "/global/common/software/desi/cori/desiconda/20200801-1.4.0-spec/conda/lib/python3.8/site-packages/pkg_resources/__init__.py", line 667, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/global/common/software/desi/cori/desiconda/20200801-1.4.0-spec/conda/lib/python3.8/site-packages/pkg_resources/__init__.py", line 1464, in run_script
    exec(code, namespace, namespace)
  File "/global/common/software/desi/cori/desiconda/20200801-1.4.0-spec/code/desispec/0.39.1/lib/python3.8/site-packages/desispec-0.39.1-py3.8.egg/EGG-INFO/scripts/desi_process_exposure", line 11, in <module>
    procexp.main(args)
  File "/global/common/software/desi/cori/desiconda/20200801-1.4.0-spec/code/desispec/0.39.1/lib/python3.8/site-packages/desispec-0.39.1-py3.8.egg/desispec/scripts/procexp.py", line 133, in main
    results, alpha = calc_tsnr2(uncalibrated_frame, fiberflat=fiberflat, skymodel=skymodel, fluxcalib=fluxcalib)
  File "/global/common/software/desi/cori/desiconda/20200801-1.4.0-spec/code/desispec/0.39.1/lib/python3.8/site-packages/desispec-0.39.1-py3.8.egg/desispec/tsnr.py", line 301, in calc_tsnr2
    log.error(f'Bad alpha={alpha:.4f} value; setting tSNR=0.0')
UnboundLocalError: local variable 'alpha' referenced before assignment

```

**Comment by sbailey** [comment]
A negative alpha is bad, but it shouldn't crash while trying to print that out.  My fault (I added the error handling code) so I will assign to myself to fix the error handling.  Too late for cascades.  I'm less certain how to debug why alpha was negative in the first place.

**Comment by michaelJwilson** [comment]
> I'm less certain how to debug why alpha was negative in the first place.
I've got code that can look at this, which I'll try to dust off within a cascades setup.  I would think it must be a pretty significant feature to pull around the value, given it's an average over the petal.  E.g. I'd think the sky model fitting must show up with something similar, if there's a DELTACHISQ to that, for instance.

**Comment by julienguy** [comment]
Fixed in PR #1164.

**Comment by sbailey** [comment]
Only 2 SV1 frames were impacted by this: 20210224-00077950-z7 and 20201222-00069401-z4 .  121 Feb/Mar 2020 frames were impacted though.

**Comment by akremin** [comment]
In the Feb and March 2020 essentially every exposure fails for z9 due to this. 

Is it a concern that this isn't random among petals? These are post-cleaning of the spectrographs early last year.
Is it obvious that z is always the problematic camera?

I spare everyone the full set, but give all the errors for two nights:

#### March 3rd: All science exposures, only camera z9.
```
poststdstar-20200303-00052925-a03679-39982481.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.3757 < 0.8
poststdstar-20200303-00052925-a03679-39982481.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052925/frame-z9-00052925.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052925/sky-z9-00052925.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052925/fluxcalib-z9-00052925.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052925/cframe-z9-00052925.fits --cosmics-nsig 6
poststdstar-20200303-00052926-a03679-39982482.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.3746 < 0.8
poststdstar-20200303-00052926-a03679-39982482.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052926/frame-z9-00052926.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052926/sky-z9-00052926.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052926/fluxcalib-z9-00052926.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052926/cframe-z9-00052926.fits --cosmics-nsig 6
poststdstar-20200303-00052927-a03679-39982483.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.3948 < 0.8
poststdstar-20200303-00052927-a03679-39982483.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052927/frame-z9-00052927.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052927/sky-z9-00052927.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052927/fluxcalib-z9-00052927.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052927/cframe-z9-00052927.fits --cosmics-nsig 6
poststdstar-20200303-00052928-a03679-39982484.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.3915 < 0.8
poststdstar-20200303-00052928-a03679-39982484.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052928/frame-z9-00052928.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052928/sky-z9-00052928.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052928/fluxcalib-z9-00052928.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052928/cframe-z9-00052928.fits --cosmics-nsig 6
poststdstar-20200303-00052932-a03679-39982485.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.3025 < 0.8
poststdstar-20200303-00052932-a03679-39982485.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052932/frame-z9-00052932.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052932/sky-z9-00052932.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052932/fluxcalib-z9-00052932.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052932/cframe-z9-00052932.fits --cosmics-nsig 6
poststdstar-20200303-00052933-a03679-39982486.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.3042 < 0.8
poststdstar-20200303-00052933-a03679-39982486.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052933/frame-z9-00052933.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052933/sky-z9-00052933.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052933/fluxcalib-z9-00052933.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052933/cframe-z9-00052933.fits --cosmics-nsig 6
poststdstar-20200303-00052934-a03679-39982487.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.3102 < 0.8
poststdstar-20200303-00052934-a03679-39982487.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052934/frame-z9-00052934.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052934/sky-z9-00052934.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052934/fluxcalib-z9-00052934.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052934/cframe-z9-00052934.fits --cosmics-nsig 6
poststdstar-20200303-00052935-a03679-39982488.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.3131 < 0.8
poststdstar-20200303-00052935-a03679-39982488.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052935/frame-z9-00052935.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052935/sky-z9-00052935.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052935/fluxcalib-z9-00052935.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052935/cframe-z9-00052935.fits --cosmics-nsig 6
poststdstar-20200303-00052936-a03679-39982489.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.3216 < 0.8
poststdstar-20200303-00052936-a03679-39982489.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052936/frame-z9-00052936.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052936/sky-z9-00052936.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052936/fluxcalib-z9-00052936.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052936/cframe-z9-00052936.fits --cosmics-nsig 6
poststdstar-20200303-00052937-a03679-39982490.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.3218 < 0.8
poststdstar-20200303-00052937-a03679-39982490.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052937/frame-z9-00052937.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052937/sky-z9-00052937.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052937/fluxcalib-z9-00052937.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052937/cframe-z9-00052937.fits --cosmics-nsig 6
poststdstar-20200303-00052938-a03679-39982491.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.3240 < 0.8
poststdstar-20200303-00052938-a03679-39982491.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052938/frame-z9-00052938.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052938/sky-z9-00052938.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052938/fluxcalib-z9-00052938.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052938/cframe-z9-00052938.fits --cosmics-nsig 6
poststdstar-20200303-00052939-a03679-39982492.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.3352 < 0.8
poststdstar-20200303-00052939-a03679-39982492.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052939/frame-z9-00052939.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052939/sky-z9-00052939.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052939/fluxcalib-z9-00052939.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052939/cframe-z9-00052939.fits --cosmics-nsig 6
poststdstar-20200303-00052940-a03679-39982493.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.3507 < 0.8
poststdstar-20200303-00052940-a03679-39982493.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052940/frame-z9-00052940.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052940/sky-z9-00052940.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052940/fluxcalib-z9-00052940.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052940/cframe-z9-00052940.fits --cosmics-nsig 6
poststdstar-20200303-00052941-a03679-39982494.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.3544 < 0.8
poststdstar-20200303-00052941-a03679-39982494.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052941/frame-z9-00052941.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052941/sky-z9-00052941.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052941/fluxcalib-z9-00052941.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052941/cframe-z9-00052941.fits --cosmics-nsig 6
poststdstar-20200303-00052942-a03679-39982495.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.4002 < 0.8
poststdstar-20200303-00052942-a03679-39982495.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052942/frame-z9-00052942.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052942/sky-z9-00052942.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052942/fluxcalib-z9-00052942.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052942/cframe-z9-00052942.fits --cosmics-nsig 6
poststdstar-20200303-00052943-a03679-39982496.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.4116 < 0.8
poststdstar-20200303-00052943-a03679-39982496.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052943/frame-z9-00052943.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052943/sky-z9-00052943.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052943/fluxcalib-z9-00052943.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052943/cframe-z9-00052943.fits --cosmics-nsig 6
poststdstar-20200303-00052961-a03679-39982497.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.3277 < 0.8
poststdstar-20200303-00052961-a03679-39982497.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052961/frame-z9-00052961.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052961/sky-z9-00052961.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052961/fluxcalib-z9-00052961.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052961/cframe-z9-00052961.fits --cosmics-nsig 6
poststdstar-20200303-00052962-a03679-39982498.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.3592 < 0.8
poststdstar-20200303-00052962-a03679-39982498.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052962/frame-z9-00052962.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052962/sky-z9-00052962.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052962/fluxcalib-z9-00052962.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052962/cframe-z9-00052962.fits --cosmics-nsig 6
poststdstar-20200303-00052978-a03679-39982499.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.4548 < 0.8
poststdstar-20200303-00052978-a03679-39982499.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052978/frame-z9-00052978.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052978/sky-z9-00052978.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052978/fluxcalib-z9-00052978.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052978/cframe-z9-00052978.fits --cosmics-nsig 6
poststdstar-20200303-00052979-a03679-39982500.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.4529 < 0.8
poststdstar-20200303-00052979-a03679-39982500.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052979/frame-z9-00052979.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052979/sky-z9-00052979.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052979/fluxcalib-z9-00052979.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052979/cframe-z9-00052979.fits --cosmics-nsig 6
poststdstar-20200303-00052993-a03679-39982501.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.5270 < 0.8
poststdstar-20200303-00052993-a03679-39982501.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052993/frame-z9-00052993.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052993/sky-z9-00052993.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052993/fluxcalib-z9-00052993.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052993/cframe-z9-00052993.fits --cosmics-nsig 6
poststdstar-20200303-00052994-a03679-39982502.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.5342 < 0.8
poststdstar-20200303-00052994-a03679-39982502.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052994/frame-z9-00052994.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052994/sky-z9-00052994.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052994/fluxcalib-z9-00052994.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052994/cframe-z9-00052994.fits --cosmics-nsig 6
poststdstar-20200303-00052995-a03679-39982503.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.5163 < 0.8
poststdstar-20200303-00052995-a03679-39982503.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052995/frame-z9-00052995.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052995/sky-z9-00052995.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052995/fluxcalib-z9-00052995.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052995/cframe-z9-00052995.fits --cosmics-nsig 6
poststdstar-20200303-00052996-a03679-39982504.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.5275 < 0.8
poststdstar-20200303-00052996-a03679-39982504.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052996/frame-z9-00052996.fits --fiberflat /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm3/fiberflatnight-z9-20200125.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052996/sky-z9-00052996.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052996/fluxcalib-z9-00052996.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200303/00052996/cframe-z9-00052996.fits --cosmics-nsig 6
```

#### March 5th, all sciences, only z9 again:
```
poststdstar-20200305-00053521-a01356789-39982506.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.4282 < 0.8
poststdstar-20200305-00053521-a01356789-39982506.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200305/00053521/frame-z9-00053521.fits --fiberflat /global/cfs/cdirs/desi/spectro/redux/cascades/calibnight/20200305/fiberflatnight-z9-20200305.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200305/00053521/sky-z9-00053521.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200305/00053521/fluxcalib-z9-00053521.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200305/00053521/cframe-z9-00053521.fits --cosmics-nsig 6
desi@cori21 [20200305]$ grep -E "ERROR|CRITICAL" *53524*log
poststdstar-20200305-00053524-a01356789-39982507.log:ERROR:tsnr.py:210:calc_alpha: tSNR alpha 0.6017 < 0.8
poststdstar-20200305-00053524-a01356789-39982507.log:CRITICAL:util.py:99:runcmd: FAILED desi_process_exposure --infile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200305/00053524/frame-z9-00053524.fits --fiberflat /global/cfs/cdirs/desi/spectro/redux/cascades/calibnight/20200305/fiberflatnight-z9-20200305.fits --sky /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200305/00053524/sky-z9-00053524.fits --calib /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200305/00053524/fluxcalib-z9-00053524.fits --outfile /global/cfs/cdirs/desi/spectro/redux/cascades/exposures/20200305/00053524/cframe-z9-00053524.fits --cosmics-nsig 6
```

**Comment by akremin** [comment]
Feb 19-Feb 24, 2020 also see basically universal failure in z2 (in addition to z9), but none after those dates. These are near the time of cleaning though, and I recall an issue in SP2, so this may be due to one of those factors impacting S/N.

**Comment by julienguy** [comment]
Yes it is a concern that this isn't random among petals. I will look at it.

**Comment by michaelJwilson** [comment]
Sorry, but I can't jump in on this today / tomorrow.

In trying to reproduce the original error, I came across that

subtract_sky
apply_fiberflat

e.g.https://github.com/desihub/desispec/blob/464708b6a46e26c51a09083ca7ce90ef2efbf5d8/py/desispec/scripts/procexp.py#L101

are changing the fiberflat and sky instances (heliocentric and fiberflat respectively) in a way that wasn't planned for in the tsnr calls.  The latter is a non-trivial effect, doubling the value of alpha.  I haven't got to think through what that means downstream.

**Comment by sbailey** [comment]
alpha<0.8 TSNR crash fixed in PR #1164, so closing this.  @michaelJwilson can you open a separate ticket explaining the fiberflat and sky issue more?  Thanks.
